src/rust/src/backend/dsa.rs

```diff
@@ -0,0 +1,333 @@
+// This file is dual licensed under the terms of the Apache License, Version
+// 2.0, and the BSD License. See the LICENSE file in the root of this repository
+// for complete details.
+
+use crate::backend::utils;
+use crate::error::{CryptographyError, CryptographyResult};
+use crate::exceptions;
+use foreign_types_shared::ForeignTypeRef;
+
+#[pyo3::prelude::pyclass(
+    module = "cryptography.hazmat.bindings._rust.openssl.dsa",
+    name = "DSAPrivateKey"
+)]
+struct DsaPrivateKey {
+    pkey: openssl::pkey::PKey<openssl::pkey::Private>,
+}
+
+#[pyo3::prelude::pyclass(
+    module = "cryptography.hazmat.bindings._rust.openssl.dsa",
+    name = "DSAPublicKey"
+)]
+struct DsaPublicKey {
+    pkey: openssl::pkey::PKey<openssl::pkey::Public>,
+}
+
+#[pyo3::prelude::pyclass(
+    module = "cryptography.hazmat.bindings._rust.openssl.dsa",
+    name = "DSAParameters"
+)]
+struct DsaParameters {
+    dsa: openssl::dsa::Dsa<openssl::pkey::Params>,
+}
+
+#[pyo3::prelude::pyfunction]
+fn private_key_from_ptr(ptr: usize) -> DsaPrivateKey {
+    let pkey = unsafe { openssl::pkey::PKeyRef::from_ptr(ptr as *mut _) };
+    DsaPrivateKey {
+        pkey: pkey.to_owned(),
+    }
+}
+
+#[pyo3::prelude::pyfunction]
+fn public_key_from_ptr(ptr: usize) -> DsaPublicKey {
+    let pkey = unsafe { openssl::pkey::PKeyRef::from_ptr(ptr as *mut _) };
+    DsaPublicKey {
+        pkey: pkey.to_owned(),
+    }
+}
+
+#[pyo3::prelude::pyfunction]
+fn generate_parameters(key_size: u32) -> CryptographyResult<DsaParameters> {
+    let dsa = openssl::dsa::Dsa::generate_params(key_size)?;
+    Ok(DsaParameters { dsa })
+}
+
+#[pyo3::prelude::pyfunction]
+fn from_private_numbers(
+    py: pyo3::Python<'_>,
+    numbers: &pyo3::PyAny,
+) -> CryptographyResult<DsaPrivateKey> {
+    let public_numbers = numbers.getattr(pyo3::intern!(py, "public_numbers"))?;
+    let parameter_numbers = public_numbers.getattr(pyo3::intern!(py, "parameter_numbers"))?;
+
+    let dsa = openssl::dsa::Dsa::from_private_components(
+        utils::py_int_to_bn(py, parameter_numbers.getattr(pyo3::intern!(py, "p"))?)?,
+        utils::py_int_to_bn(py, parameter_numbers.getattr(pyo3::intern!(py, "q"))?)?,
+        utils::py_int_to_bn(py, parameter_numbers.getattr(pyo3::intern!(py, "g"))?)?,
+        utils::py_int_to_bn(py, numbers.getattr(pyo3::intern!(py, "x"))?)?,
+        utils::py_int_to_bn(py, public_numbers.getattr(pyo3::intern!(py, "y"))?)?,
+    )
+    .unwrap();
```

Comment:
Why unwrap? We handle `ErrorStack` conversion to `CryptographyError` already so should we be doing `?`?

Reply:
Two parter:

1) The `?` is uncovered if I use it, because...
2) These _cannot fail_, I reviewed both rust-openssl and OpenSSL itself. There is no way for it to fail.